### ISSUE TITLE
A Better Service Experience

### ISSUE THREAD

**Issue by anthonylavado** [jellyfin/jellyfin-server-windows#29]
Even though using Jellyfin as a Windows service is **not recommended**, we should take steps to improve this to be more robust for the future.

## Improve the Service Host
- [ ] Use the tray app as a foundation for a new Service Host app, removing NSSM
  - [ ] New service host should start the process in the background (no EditBin hack needed)
### Temporary Fixes
- [x] Use correct format for Network Service account in NSSM (#39)

## Improve Logging

- [ ] Write to Windows Event log in any error case

---

This is part of a larger restructuring effort to improve the experience on Windows.

**Comment by MrGrymReaper** [comment]
It has extensions functionality which allow for it to be extended in many different ways, including a providing a tray icon.

It also planning several enhancement features which will improve security of the wrapped service programs. Also using a tray icon as a service host isn't recommended for several Windows environments, they tend to get disabled as their launching slows system operating down. Plus they only start when user is logged on, however if they are started from WinSW via an extension then it may not be as bad.

It has deferred file operation functionality which can be of extreme use to Jellyfin. This functionality allows for the possibility to defer the updating of files, while a wrapped program is running.

**Comment by MrGrymReaper** [comment]
It also causes my anti-malware system to eventually evaluate the impact of the program running from a shortcut in startup group, or from registry in startup and operating speed. Running as actual Windows Service isn't evaluated as it doesn't have an impact on startup speed, also they run when user isn't logged on.

**Comment by MrGrymReaper** [comment]
Your current idea of using tray icons also can potentially be a source of problems if running on a host environment which is based around Windows Server Core (e.g. Windows Server 2012 Core, Windows Server 2012 R2 Core, Windows Server 2016 Core, Windows Server 2019 Core etc). As there's no tray area under this environment due not having a full desktop environment install currently.

**Comment by anthonylavado** [comment]
> It has extensions functionality which allow for it to be extended in many different ways, including a providing a tray icon.

I'm sorry, what are you referring to here?

I'll address the other comments shortly.

**Comment by MrGrymReaper** [comment]
> > It has extensions functionality which allow for it to be extended in many different ways, including a providing a tray icon.
> 
> I'm sorry, what are you referring to here?
> 
> I'll address the other comments shortly.

I mean you can extend the functionality of WinSW with extensions as needed to add any needed functionality.

**Comment by anthonylavado** [comment]
@MrGrymReaper Okay, I see now that you are referring to [WinSW](https://github.com/winsw/winsw). The way your comment was phrased didn't make this clear at first.

Before we continue, let me explain my comment further. In the initial issue post:
>Even though using Jellyfin as a Windows service is **not recommended**

Right now we're the unwilling victims of an FFmpeg bug where if FFmpeg is called by an application without access to a desktop environment, it will not be able to run with Hardware Acceleration. For reference, see:

- [FFmpeg Windows version with QSV hwaccel fails over TERMINAL](https://trac.ffmpeg.org/ticket/7511)
- [Intel QSV: "Failed to create Direct3D device" on Core i7-7700K (Skylake) on Windows 10](https://trac.ffmpeg.org/ticket/6827)
- [Direct3D 9 is not supported in the context of a Windows service](https://support.microsoft.com/en-us/help/978635/direct3d-9-is-not-supported-in-the-context-of-a-windows-service)



>It is also planning several enhancement features which will improve security of the wrapped service programs.

Could you highlight what would be beneficial here? On the one hand, we want to move _away_ from external dependencies as much as we can.

>Also using a tray icon as a service host isn't recommended for several Windows environments...

Apologies, I wasn't clear here. I'm not saying we will have a tray icon that launches the service, I want to use the _code_ of the app to write a utility that launches Jellyfin in the background, to be used as a service. In practice, that's basically just these lines:
```csharp
Process p = new Process();
p.StartInfo.FileName = _executableFile;
p.StartInfo.CreateNoWindow = true;
p.StartInfo.Arguments = "--noautorunwebapp --datadir \"" + _dataFolder + "\"";
p.Start();
```

>... tend to get disabled... ...only start when user is logged on...

That's the goal here, to make sure it starts in all occasions in a reliable manner. We had thought of some alternatives before (https://github.com/jellyfin/jellyfin/issues/1953), but as I've said, I want to get away from hacks and having to ship two copies of the executable (one that launches silently and one that doesn't). If we control the "launcher", then we can also have it share frameworks that are already in place (see https://github.com/jellyfin/jellyfin-server-windows/issues/28). If you look at how WinSW is built ([WinSW build.yml](https://github.com/winsw/winsw/blob/0bdd21b6c0077e05f78369135a9f7283df61cd76/eng/build.yml#L34)), you'll see they're using the same options we did when building the tray app (ref [.csproj](https://github.com/jellyfin/jellyfin-server-windows/blob/5878dc7e50dd389e96f41279ee9146b3e9441c0f/Jellyfin.Windows.Tray/Jellyfin.Windows.Tray.csproj#L9-L10)). We've had so many issues with Publish Single File, that I've had to make a dedicated issue about it (see https://github.com/jellyfin/jellyfin-server-windows/issues/27, and I would link all the other issues and threads, but it's not worth it).

> ...inbuilt support for mapping SMB shares or other shares...

Is this referring to [sharedDirectoryMapper](https://github.com/winsw/winsw/blob/master/doc/extensions/sharedDirectoryMapper.md), and https://github.com/winsw/winsw/issues/538? I will agree that getting shares accessible by the account running the service is tricky, but again, I don't know about introducing a whole dependency for this. I would also want the concerns in 538 to be addressed (and https://github.com/winsw/winsw/issues/439) for that matter.


>It has deferred file operation functionality which can be of extreme use to Jellyfin.
Is this about automatic updating for Jellyfin? I want to make it *super* optional, and I've got another thread to talk about it: https://github.com/jellyfin/jellyfin-server-windows/issues/30. The key point would be to have a notification first (which would probably be of best use in the server, with a notification/banner in the dashboard). Haven't thought about the mechanism to do it with yet. The automatic updating of WinSW, though potentially useful, would mean I'd have to recreate the config file based on settings, from the server, and potentially re-install the service. Not great.

>...nsis installer can easily generate the required config file... ... run the appropriate service install action...

Yes, but I also can run commands to do this directly with Windows, bypassing WinSW.

>...running from a shortcut in startup group... / ...from registry in startup... / ...Running as actual Windows Service...  ...run when user isn't logged on.

It's not, but you knew that. / That's [how it's done now](https://github.com/jellyfin/jellyfin-server-windows/blob/5878dc7e50dd389e96f41279ee9146b3e9441c0f/Jellyfin.Windows.Tray/TrayApplicationContext.cs#L234-L252), which is why I want to improve this. **But**, that's only for the tray icon when *not* in service mode. / That's the goal, and how our current NSSM solution works now (though it's still not great).

>...source of problems if running on a host environment which is based around Windows Server Core...

As per above, I'm not planning to run a tray icon to run the service.

I'm open to ideas, but hopefully this explains some of the motivation behind this, and also provides some understanding of the challenges we face making a software program that is both powerful, and developed in our spare time 😄

**Comment by MrGrymReaper** [comment]
What version of ffmpeg is being used currently as a base for your build? Is it version of ffmpeg 4.1 or newer (preferably 4.2 or a spin off of the git) as the issue may have, been fixed there. Which you reference to in regards to DirectX 9 as practically every fairly new device's GPU now has at least DirectX 11 support present.

**Comment by anthonylavado** [comment]
> What version of ffmpeg is being used currently as a base for your build?

We currently ship a customized FFmpeg for Windows, because it needed to include a fix for segment muxing, which we haven't been able to get upstream yet.

>Is it version of ffmpeg 4.1 or newer (preferably 4.2 or a spin off of the git)...

It's 4.2.1.

>as the issue may have, been fixed there.

It hasn't. See https://trac.ffmpeg.org/ticket/7933 (which has a possible solution in comment 3).

>Which you reference to in regards to DirectX 9 as practically every fairly new device's GPU now has at least DirectX 11 support present.

GPUs do, but FFmpeg doesn't do that in Intel QSV init, which is the problem. See https://trac.ffmpeg.org/ticket/7511 again, specifically comments 8 & 9.

**Comment by makuc** [comment]
Manually copying over the following DDLs (from the Windows Server Installation Media, Sources\install.wim file opened in 7-Zip,  check in various single-number folders representing different versions until you locate the one with Desktop Experience, mine was 4) makes the Jellyfin Server work properly on Windows Server Core (2022 Insider Preview) installed as a service. Some might not necessarily be needed, but since it works now, I have no intention of experimenting anymore.

System32 and SysWOW64 folders, each its own copies from the installation media (above):
- avicap32.dll
- ddraw.dll
- dsound.dll
- dxva2.dll
- glu32.dll
- mfplat.dll
- MFPlay.dll
- msvfw32.dll
- opengl32.dll

**Comment by bilbospocketses** [comment]
Hi @anthonylavado  

I am following up on my suggestions from discord back in April:

[Discord Dev Channel](https://discord.com/channels/772232779534172171/772247115681890314/1498104411845693605)

Here's the high-level plan I've pulled together. I re-read this issue (plus #27, #28, #30) so it builds on what you've already laid out instead of reinventing it.

The goal is to create a Windows service that starts reliably and doesn't break anyone's setups on upgrade. The mentality is that it is better to own the launcher (out of the tray code) than keep leaning on nssm, and skip the EditBin / two-copies-of-the-exe hacks.

Plan would be to work through it in two phases:

Phase 1 — modernize the tray. Get `Jellyfin.Windows.Tray` from net472 onto current .NET (server's on net10 now), with no behavior changes. Just get the codebase on a modern runtime so the service has something clean to sit on.

For packaging: no single-file — either self-contained as a folder, or framework-dependent sharing one runtime with the rest of the install.

Phase 2 — drop nssm using something similar to what you had suggested (`jellyfin.exe --noautorunwebapp --datadir …`, no window), wrapped as a real Windows service via the modern .NET stack, plus the Event Log writes on errors.

The biggest concern is migration. The plan would be for the installer to detect the existing setup, carry over the data dir / account / config, hand off, and pull nssm cleanly, with a fallback if anything looks off. A lot of testing needed for this part.

And while we're in there: once the launcher is finished, an opt-in auto-updater (#30) gets a lot easier to implement as part of the deployment. The service can coordinate the stop / swap / restart cleanly instead of fighting nssm. Probably best to keep it opt-in and notify-first like you mentioned. The actual "there's an update" notification likely still belongs in the server dashboard rather than the tray (or we can swap in a different tray icon when there's an update, or even better - both, with a toast on startup, mix and match, whatever you think)

Where I could use your guidance:

1. net10 to match the server, or net8 LTS for now?
2. packaging — self-contained folder or framework-dependent / shared runtime?
3. service as a `--service` mode of the tray, or a separate project in the repo?
4. auto-update (#30) — does opt-in + notify-first sound right, and where should update notifications live (toast on startup, dashboard, tray icon swap)?
5. anything in the original issue that's gone stale, or anything else that would be nice to roll in?

Let me know what you think whenever you have a chance. Happy to kick off with Phase 1 as a small PR anytime.

**Comment by thornbill** [comment]
@bilbospocketses Try reading the llm policy first.

**Comment by bilbospocketses** [comment]
Hi @thornbill 

Thanks for the note. I've read the policy and am definitely aware of folks creating junk code with AI tools. For sure that is not me. I am way too anal retentive for that.

**Comment by thornbill** [comment]
Just a heads up anytime your slop machine references a PR github very helpfully links back to it 😉

https://github.com/bilbospocketses/jellyfin-server-windows/blob/windows-service-tray/docs/superpowers/plans/2026-06-06-jellyfin-windows-tray-net10-modernization.md#wave-0--engage-on-issue-29-do-this-first
